### PR TITLE
DomainParticipant get/set_default_subscriber_qos

### PR DESCRIPTION
**Pull request by IkerLuengo**
- Implementation of get/set_default_subscriber_qos on PIM
- Unit test for get/set_default_subscriber_qos on PIM
- link PSM get/set_default_subscriber_qos methods to PIM
- Unit test for get/set_default_subscriber_qos on PSM

NOTE: This PR must be merged after #1073

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1740/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1497/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1732/)

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 1]
```diff
@@ -343,22 +344,22 @@ const fastdds::dds::PublisherQos& DomainParticipantImpl::get_default_publisher_q
...
     {
-        default_sub_qos_.set_qos(SUBSCRIBER_QOS_DEFAULT, true);
+        SubscriberImpl::set_qos(default_sub_qos_, SUBSCRIBER_QOS_DEFAULT, true);
         return ReturnCode_t::RETCODE_OK;
     }
-    else if (qos.check_qos())
+    else if (SubscriberImpl::check_qos(qos))
     {
-        default_sub_qos_.set_qos(qos, false);
+        SubscriberImpl::set_qos(default_sub_qos_, qos, false);
```
Comment (laura-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:356]:
```suggestion
        SubscriberImpl::set_qos(default_sub_qos_, qos, true);
```

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 2]
```diff
@@ -343,22 +344,22 @@ const fastdds::dds::PublisherQos& DomainParticipantImpl::get_default_publisher_q
...
 ReturnCode_t DomainParticipantImpl::set_default_subscriber_qos(
-        const fastdds::dds::SubscriberQos& qos)
+        const SubscriberQos& qos)
 {
     if (&qos == &SUBSCRIBER_QOS_DEFAULT)
     {
-        default_sub_qos_.set_qos(SUBSCRIBER_QOS_DEFAULT, true);
+        SubscriberImpl::set_qos(default_sub_qos_, SUBSCRIBER_QOS_DEFAULT, true);
         return ReturnCode_t::RETCODE_OK;
     }
```
Comment (laura-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:353]: It is not necessary to check if qos is equal to SUBSCRIBER_QOS_DEFAULT

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 3]
```diff
@@ -343,22 +344,22 @@ const fastdds::dds::PublisherQos& DomainParticipantImpl::get_default_publisher_q
...
-    else if (qos.check_qos())
+    else if (SubscriberImpl::check_qos(qos))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:354]: Check also if the qos can be updated

File: src/cpp/fastdds/subscriber/SubscriberImpl.cpp [hunk 4]
```diff
@@ -90,13 +90,13 @@ const SubscriberQos& SubscriberImpl::get_qos() const
 ReturnCode_t SubscriberImpl::set_qos(
         const SubscriberQos& qos)
 {
-    if (!qos.check_qos())
+    if (!check_qos(qos))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.cpp:93]:
```suggestion
    if (check_qos(qos))
```

File: src/cpp/fastdds/subscriber/SubscriberImpl.cpp [hunk 5]
```diff
@@ -460,6 +460,50 @@ bool SubscriberImpl::type_in_use(
...
+    }
+    if (to.group_data().getValue() != from.group_data().getValue() )
+    {
+        to.group_data() = from.group_data();
+        to.group_data().hasChanged = true;
+    }
+    if (to.entity_factory().autoenable_created_entities != from.entity_factory().autoenable_created_entities)
+    {
+        to.entity_factory() = from.entity_factory();
+        to.entity_factory().hasChanged = true;
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.cpp:487]:
```suggestion
```
As this Qos is not going to be sent it is not necessary to set hasChanged to true

File: src/cpp/dds/domain/DomainParticipant.cpp [hunk 6]
```diff
@@ -133,17 +133,21 @@ DomainParticipant& DomainParticipant::default_publisher_qos(
...
-//    return *this;
-//}
+DomainParticipant& DomainParticipant::default_subscriber_qos(
+        const ::dds::sub::qos::SubscriberQos& qos)
+{
+    if (this->delegate()->set_default_subscriber_qos(qos) ==
+            ReturnCode_t::RETCODE_INCONSISTENT_POLICY)
+    {
+        throw dds::core::InconsistentPolicyError("Inconsistent Qos");
+    }
```
Comment (laura-eprosima) [inline src/cpp/dds/domain/DomainParticipant.cpp:148]: Check also if the ReturnCode is RETCODE_IMMUTABLE_POLICY

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp (hunk 1 above)
Comment (IkerLuengo) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:356]: :+1:

File: src/cpp/fastdds/subscriber/SubscriberImpl.cpp (hunk 5 above)
Comment (IkerLuengo) [reply src/cpp/fastdds/subscriber/SubscriberImpl.cpp]: Then it probably doesn't need to iinherit from QosPolicy either. I'm changing that too.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1841/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1587/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1835/)

File: src/cpp/fastdds/subscriber/SubscriberImpl.cpp [hunk 7]
```diff
@@ -90,16 +90,17 @@ const SubscriberQos& SubscriberImpl::get_qos() const
 ReturnCode_t SubscriberImpl::set_qos(
         const SubscriberQos& qos)
 {
-    if (!qos.check_qos())
+    ReturnCode_t check_result = SubscriberImpl::check_qos(qos);
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.cpp:93]: Check also if SUBSCRIBER_QOS_DEFAULT
```suggestion
    if (&qos == &SUBSCRIBER_QOS_DEFAULT)
    {
        const SubscriberQos& default_qos = participant_->get_default_subscriber_qos();
        if (!can_qos_be_updated(qos_, default_qos))
        {
            return ReturnCode_t::RETCODE_IMMUTABLE_POLICY;
        }

        set_qos(qos_, default_qos, false);
        return ReturnCode_t::RETCODE_OK;
    }
    ReturnCode_t check_result = SubscriberImpl::check_qos(qos);
```

File: src/cpp/dds/domain/DomainParticipant.cpp [hunk 8]
```diff
@@ -133,17 +133,25 @@ DomainParticipant& DomainParticipant::default_publisher_qos(
...
+{
+    ReturnCode_t result = delegate()->set_default_subscriber_qos(qos);
+    if (result == ReturnCode_t::RETCODE_INCONSISTENT_POLICY)
+    {
+        throw dds::core::InconsistentPolicyError("Inconsistent Qos");
+    }
+    if (result == ReturnCode_t::RETCODE_UNSUPPORTED)
+    {
+        throw dds::core::UnsupportedError("Unsupported Qos");
+    }
```
Comment (laura-eprosima) [inline src/cpp/dds/domain/DomainParticipant.cpp:152]: This check is not necessary as never returns UNSUPPORTED. Also, change RETCODE_INCONSISTENT_POLICY by RETCODE_IMMUTABLE_POLICY as this is the only one that can be returned apart from OK.

**Comment by IkerLuengo** [comment]
Closed as the changes are already included in #1122

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1854/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1596/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1848/)

**Pull request closed without merging**